File: apps/web/src/lib/task-progress.ts
type SubtaskProgressItem = {
  is_done: boolean;
};

export type SubtaskProgressSummary = {
  total: number;
  done: number;
  percent: number | null;
};

export function getSubtaskProgressSummary(subtasks: readonly SubtaskProgressItem[]): SubtaskProgressSummary {
  const total = subtasks.length;

  if (total === 0) {
    return {
      total: 0,
      done: 0,
      percent: null,
    };
  }

  const done = subtasks.reduce((count, subtask) => count + (subtask.is_done ? 1 : 0), 0);

  return {
    total,
    done,
    percent: Math.round((done / total) * 100),
  };
}

File: apps/android/app/src/main/java/com/taskmanager/android/domain/TaskProgress.kt
package com.taskmanager.android.domain

import kotlin.math.roundToInt

data class SubtaskProgressSummary(
    val done: Int,
    val total: Int,
    val percent: Int?,
)

fun getSubtaskProgressSummary(done: Int, total: Int): SubtaskProgressSummary = SubtaskProgressSummary(
    done = done,
    total = total,
    percent = if (total == 0) null else ((done.toFloat() / total.toFloat()) * 100f).roundToInt(),
)
